python/GafferUI/AnimationEditor.py

```diff
@@ -311,4 +317,309 @@ def __repr__( self ) :
+			with Gaffer.BlockedConnection( self.__frameConnection ) :
+				self.__frameEditor.setText( "" )
+				self.__frameEditor._qtWidget().setPlaceholderText( "---" )
+
+		# set enabled when all selected keys have different parent curves
+		enabled = bool( selectedKeys )
+		curves = set()
+		for key in selectedKeys :
+			if key.parent() in curves :
+				enabled = False
+				break
+			curves.add( key.parent() )
+		self.__frameEditor.setEnabled( enabled )
+
+	def __updateKeyValue( self ) :
+
+		# if multiple keys selected display "---" unless all selected keys have same value
+		selectedKeys = self.parent().curveGadget().selectedKeys()
+		value = None
+		if len( selectedKeys ) > 1 :
+			value = selectedKeys[ 0 ].getValue()
+			for key in selectedKeys[1:] :
+				if value != key.getValue() :
+					value = None
+					break
+		elif selectedKeys :
+			value = selectedKeys[ 0 ].getValue()
+		if value is not None :
+			with Gaffer.BlockedConnection( self.__valueConnection ) :
+				self.__valueEditor.setValue( value )
+		else :
+			with Gaffer.BlockedConnection( self.__valueConnection ) :
+				self.__valueEditor.setText( "" )
+				self.__valueEditor._qtWidget().setPlaceholderText( "---" )
+
+		# set disabled when no selected keys
+		self.__valueEditor.setEnabled( bool( selectedKeys ) )
+
+	def __updateKeyInterpolation( self ) :
+
+		mode = self.getInterpolationForSelectedKeys()
+		self.__interpolationEditor.setText( "---" if mode is None else mode.name )
+
+		# set disabled when no selected keys
+		self.__interpolationEditor.setEnabled( bool( self.parent().curveGadget().selectedKeys() ) )
+
+	def __setKeyInterpolation( self, unused, mode ) :
+
+		# set interpolation for all selected keys
+		selectedKeys = self.parent().curveGadget().selectedKeys()
+		if selectedKeys :
+			with Gaffer.UndoScope( selectedKeys[0].parent().ancestor( Gaffer.ScriptNode ) ) :
+				for key in selectedKeys :
+					with Gaffer.BlockedConnection( self.__connections[ key.parent() ].interpolation ) :
+						key.setInterpolation( mode )
+		self.__interpolationEditor.setText( mode.name )
+
+	def __setFrame( self, widget, reason ) :
+
+		# check for invalid edit
+		if reason == GafferUI.NumericWidget.ValueChangedReason.InvalidEdit :
+			self.__updateKeyFrame()
+			return
+
+		# handle undo queue
+		if not widget.changesShouldBeMerged( self.__lastChangedReasonFrame, reason ) :
+			curves = self.parent().curveGadget().editablePlugs()
+			if curves :
+				scriptNode = curves[0].ancestor( Gaffer.ScriptNode )
+				with Gaffer.UndoScope( scriptNode, mergeGroup=str( self.__mergeGroupIdFrame ) ) :
+					for curve in curves :
+						curve.removeInactiveKeys()
+			self.__mergeGroupIdFrame += 1
+		self.__lastChangedReasonFrame = reason
+
+		# only set frame for multiple keys if all selected keys have different parent curves
+		# as setting two keys on the same curve to the same frame will delete all but one of the keys.
+		selectedKeys = self.parent().curveGadget().selectedKeys()
+		if selectedKeys :
```

Comment:
There are two sources of `valueChangedSignal()` emission : the user interacting with the widget, and a call to `NumericWidget.setValue()`. `setEnabled( False )` prevents the first, and a BlockedConnection prevents the second reaching us when we call `__frameEditor.setValue( frame )`. Nobody else can call `setValue()` because this is a private widget. Maybe I missed something? But if not, it'd be good to avoid repeating this logic...